lib/FunctionIsolation/IsolatedFunctionKind.cpp

```diff
@@ -0,0 +1,184 @@
+        auto Function = std::get<model::Function>(*Change.Old);
+        insertCallers(CrossRelations, Function.Entry, Targets);
+      } else {
+        revng_abort();
+      }
+    }
+
+    // A new `model::DynamicFunction` has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/ImportedDynamicFunctions") {
+      if (Change.New) {
+        // Invalidate all targets
+        invalidateAllTargetsPerFunctionRank(Model, Targets);
+      } else if (Change.Old) {
+        // Invalidate model::DynamicFunction's callers
+        auto DynamicFunction = std::get<model::DynamicFunction>(*Change.Old);
+        insertCallers(CrossRelations, DynamicFunction.OriginalName, Targets);
+      } else {
+        revng_abort();
+      }
+    }
+
+    // Did FinalStackOffset or PreservedRegisters for a `model::Function` have
+    // changed?
+    // Note that, since matching e.g., FinalStackOffset, for both
+    // RawFunctionType and CABIFunctionType is non-trivial, we will not leverage
+    // the diff for checking the change related to FSO. Instead, we move to
+    // `Layout` by employing the old model, and the new one (the one with the
+    // change).
+    for (const model::Function &Function : Model->Functions) {
+      MetaAddress Entry = Function.Entry;
+      if (OldModel->Functions.find(Entry) != OldModel->Functions.end())
+        if (haveFSOAndPreservedRegsChanged(OldModel->Functions.at(Entry)
+                                             .Prototype,
+                                           Function.Prototype)) {
+          Targets.insert(Function.Entry);
+          insertCallersAndTransitiveClosureIfInline(CrossRelations,
+                                                    Model,
+                                                    Entry,
+                                                    Targets);
+        }
+    }
+
+    // Did FinalStackOffset or PreservedRegisters for a `model::DynamicFunction`
+    // have changed?
+    for (const model::DynamicFunction &Function : IDF) {
+      std::string Name = Function.OriginalName;
+      const auto &OldIDF = OldModel->ImportedDynamicFunctions;
+      if (OldIDF.find(Name) != OldIDF.end()) {
+        if (haveFSOAndPreservedRegsChanged(getPrototype(*OldModel,
+                                                        OldIDF.at(Name)),
+                                           getPrototype(*Model, Function)))
+          insertCallersAndTransitiveClosureIfInline(CrossRelations,
+                                                    Model,
+                                                    Name,
+                                                    Targets);
+      }
+    }
+
+    // Did attributes for a `model::Function` have changed?
+    auto MaybeFuncAttr = FuncAttrMatcher.match<MetaAddress>(Change.Path);
+    if (MaybeFuncAttr) {
+      auto Entry = std::get<MetaAddress>(*MaybeFuncAttr);
+      auto NewAttribute = std::get<FunctionAttribute>(*Change.New);
```

Comment:
You're assuming attributes can only be removed, it seems.

Reply:
You mean I'm assuming attributes can only be added?

Reply:
Yep